### PR TITLE
fix(lsp): valid bufnr befor save

### LINKED ISSUE #23885: Codelens refresh causes unending stacktrace when deleting a buffer before lenses are resolved
### Problem

When vim.lsp.codelens.refresh is invoked for a buffer, but the buffer is closed before the codelenses can be resolved i.e., the codelens is still in the `Unresolved codelens...`  state, an error is thrown as shown in the screenshot and on pressing `<enter>` the stacktrace just keeps popping up.

Screenshot 2023-06-02 at 1 55 28 PM

### Steps to reproduce using "nvim -u minimal_init.lua"

```lua
--- CHANGE THESE
local pattern = 'java'
local cmd = {'jdtls'}
-- Add files/folders here that indicate the root of a project
local root_markers = {'.git', 'mvnw', 'gradlew'}
-- Change to table with settings if required
local settings = {
    java = {
      signatureHelp = { enabled = true },
      contentProvider = { preferred = 'fernflower' },
      completion = {
        matchCase = 'FIRSTLETTER',
        guessMethodArguments = true,
        favoriteStaticMembers = {
          "org.hamcrest.MatcherAssert.assertThat",
          "org.hamcrest.Matchers.*",
          "org.hamcrest.CoreMatchers.*",
          "org.junit.jupiter.api.Assertions.*",
          "java.util.Objects.requireNonNull",
          "java.util.Objects.requireNonNullElse",
          "java.util.AbstractMap.SimpleEntry",
          "java.util.AbstractMap.SimpleImmutableEntry",
          "org.mockito.Mockito.*",
          "org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*",
          "org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*",
          "org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*",
          "org.springframework.test.web.servlet.result.MockMvcResultMatchers.*",
          "org.springframework.test.web.servlet.result.MockMvcResultMatchers.*",
        },
        importOrder = {
          "#",
          ""
        }
      },
      sources = {
        organizeImports = {
          starThreshold = 9999,
          staticStarThreshold = 9999,
        },
      },
      codeGeneration = {
        toString = {
          template = "${object.className}{${member.name()}=${member.value}, ${otherMembers}}"
        }
      },
      maven = {
        downloadSources = true,
        updateSnapshots = true
      },
      implementationsCodeLens = {
        enabled = true,
      },
      referencesCodeLens = {
        enabled = true,
      },
      references = {
        includeDecompiledSources = true,
      },
      inlayHints = {
        parameterNames = {
          enabled = true,
        },
      },
      format = {
        enabled = false,
      },
      configuration = {
        updateBuildConfiguration = "interactive"
      }
    }
  }

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'jdtls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings
    })
  end
})
```

### Expected behavior

Expect no stack trace to be thrown on buffer delete before codelens refresh

### Neovim version (nvim -v)

NVIM v0.9.0

### Language server name/version

jdtls 1.23.0

### Operating system/version

macos big sur

### Log file

_No response_

### PR DESCRIPTION
**Pull request by snowair**
fixes https://github.com/neovim/neovim/issues/23885

### CODE REVIEW

**Comment by justinmk** [comment]
What is calling these handlers? Should it validate the buffer instead of blindly calling handlers that expect a buffer?

**Comment by glepnir** [comment]
it's better in `on_codelens` handler . valid `ctx.bufnr`.

File: runtime/lua/vim/lsp/codelens.lua [hunk 1]
```diff
@@ -199,6 +206,9 @@ end
 
 ---@private
 local function resolve_lenses(lenses, bufnr, client_id, callback)
+  if not api.nvim_buf_is_valid(bufnr) then
+    return
+  end
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/codelens.lua:211]: I think this still needs to call the callback to ensure the `active_refreshes`  table entry is removed

**Comment by mfussenegger** [comment]
https://github.com/neovim/neovim/pull/23887 included a variant of the changes here.

**Pull request closed without merging**